File: backend/services/review_service.py
from backend.extensions import db 
from sqlalchemy.exc import SQLAlchemyError
from backend.models import BakeryReview, ProductReview 
class ReviewService:

    """Service class for review-related business logic"""
    
    # === Bakery Review Methods ===
    
    def get_all_bakery_reviews(self):
        """Get all bakery reviews ordered by creation date (newest first)"""
        return BakeryReview.query.order_by(BakeryReview.created_at.desc()).all()
    
    def get_bakery_review_by_id(self, review_id):
        """Get a specific bakery review by ID"""
        return BakeryReview.query.get(review_id)
    
    def get_bakery_reviews_by_bakery(self, bakery_id):
        """Get all reviews for a specific bakery"""
        return BakeryReview.query.filter_by(bakery_id=bakery_id).order_by(BakeryReview.created_at.desc()).all()
    
    def get_bakery_reviews_by_user(self, user_id):
        """Get all bakery reviews by a specific user"""
        return BakeryReview.query.filter_by(user_id=user_id).order_by(BakeryReview.created_at.desc()).all()
    
    def create_bakery_review(self, review, overall_rating, service_rating, price_rating, 
                         atmosphere_rating, location_rating, user_id=None, bakery_id=None):
        """Create a new bakery review - user_id now optional"""
        try:
            new_review = BakeryReview(
                review=review,
                overall_rating=overall_rating,
                service_rating=service_rating,
                price_rating=price_rating,
                atmosphere_rating=atmosphere_rating,
                location_rating=location_rating,
                user_id=user_id,  # Can be None for anonymous reviews
                bakery_id=bakery_id
            )
            db.session.add(new_review)
            db.session.commit()
            return new_review
        except SQLAlchemyError as e:
            db.session.rollback()
            raise Exception(f"Database error: {str(e)}")
    
    def update_bakery_review(self, review_id, review, overall_rating, service_rating, price_rating, 
                         atmosphere_rating, location_rating, user_id=None, bakery_id=None):
        """Update an existing bakery review - user_id now optional"""
        try:
            bakery_review = self.get_bakery_review_by_id(review_id)
            if not bakery_review:
                raise Exception("Bakery review not found")
                
            bakery_review.review = review
            bakery_review.overall_rating = overall_rating
            bakery_review.service_rating = service_rating
            bakery_review.price_rating = price_rating
            bakery_review.atmosphere_rating = atmosphere_rating
            bakery_review.location_rating = location_rating
            bakery_review.user_id = user_id  # Can be None for anonymous reviews
            bakery_review.bakery_id = bakery_id
            
            db.session.commit()
            return bakery_review
        except SQLAlchemyError as e:
            db.session.rollback()
            raise Exception(f"Database error: {str(e)}")
    
    def delete_bakery_review(self, review_id):
        """Delete a bakery review"""
        try:
            bakery_review = self.get_bakery_review_by_id(review_id)
            if not bakery_review:
                raise Exception("Bakery review not found")
                
            db.session.delete(bakery_review)
            db.session.commit()
            return True
        except SQLAlchemyError as e:
            db.session.rollback()
            raise Exception(f"Database error: {str(e)}")
    
    # === Product Review Methods ===
    
    def get_all_product_reviews(self):
        """Get all product reviews ordered by creation date (newest first)"""
        return ProductReview.query.order_by(ProductReview.created_at.desc()).all()
    
    def get_product_review_by_id(self, review_id):
        """Get a specific product review by ID"""
        return ProductReview.query.get(review_id)
    
    def get_product_reviews_by_product(self, product_id):
        """Get all reviews for a specific product"""
        return ProductReview.query.filter_by(product_id=product_id).order_by(ProductReview.created_at.desc()).all()
    
    def get_product_reviews_by_user(self, user_id):
        """Get all product reviews by a specific user"""
        return ProductReview.query.filter_by(user_id=user_id).order_by(ProductReview.created_at.desc()).all()
    
    def create_product_review(self, review, overall_rating, taste_rating, price_rating, 
                         presentation_rating, user_id=None, product_id=None):
        """Create a new product review - user_id now optional"""
        try:
            new_review = ProductReview(
                review=review,
                overall_rating=overall_rating,
                taste_rating=taste_rating,
                price_rating=price_rating,
                presentation_rating=presentation_rating,
                user_id=user_id,  # Can be None for anonymous reviews
                product_id=product_id
            )
            db.session.add(new_review)
            db.session.commit()
            return new_review
        except SQLAlchemyError as e:
            db.session.rollback()
            raise Exception(f"Database error: {str(e)}")
    
    def update_product_review(self, review_id, review, overall_rating, taste_rating, price_rating, 
                         presentation_rating, user_id=None, product_id=None):
        """Update an existing product review - user_id now optional"""
        try:
            product_review = self.get_product_review_by_id(review_id)
            if not product_review:
                raise Exception("Product review not found")
                
            product_review.review = review
            product_review.overall_rating = overall_rating
            product_review.taste_rating = taste_rating
            product_review.price_rating = price_rating
            product_review.presentation_rating = presentation_rating
            product_review.user_id = user_id  # Can be None for anonymous reviews
            product_review.product_id = product_id
            
            db.session.commit()
            return product_review
        except SQLAlchemyError as e:
            db.session.rollback()
            raise Exception(f"Database error: {str(e)}")
    
    def delete_product_review(self, review_id):
        """Delete a product review"""
        try:
            product_review = self.get_product_review_by_id(review_id)
            if not product_review:
                raise Exception("Product review not found")
                
            db.session.delete(product_review)
            db.session.commit()
            return True
        except SQLAlchemyError as e:
            db.session.rollback()
            raise Exception(f"Database error: {str(e)}")

File: frontend/src/services/reviewService.js
import { BaseService } from './baseService';
import apiClient from './api';

class ReviewService extends BaseService {
  constructor() {
    super('');
  }

  // Bakery Reviews
  async getAllBakeryReviews() {
    try {
      const response = await apiClient.get('/bakeryreviews', true);
      return response;
    } catch (error) {
      console.error('Error fetching bakery reviews:', error);
      throw error;
    }
  }

  async getBakeryReviewsByUser(userId) {
    try {
      const response = await apiClient.get(`/bakeryreviews/user/${userId}`, true);
      return response;
    } catch (error) {
      console.error(`Error fetching bakery reviews for user ${userId}:`, error);
      throw error;
    }
  }
  
  async getBakeryReviewsByBakery(bakeryId) {
    try {
      const response = await apiClient.get(`/bakeryreviews/bakery/${bakeryId}`, true);
      return response;
    } catch (error) {
      console.error(`Error fetching reviews for bakery ${bakeryId}:`, error);
      throw error;
    }
  }

  async createBakeryReview(reviewData) {
    try {
      const response = await apiClient.post('/bakeryreviews/create', reviewData);
      return response;
    } catch (error) {
      console.error('Error creating bakery review:', error);
      throw error;
    }
  }

  async updateBakeryReview(id, reviewData) {
    try {
      const response = await apiClient.patch(`/bakeryreviews/update/${id}`, reviewData);
      return response;
    } catch (error) {
      console.error(`Error updating bakery review ${id}:`, error);
      throw error;
    }
  }

  async deleteBakeryReview(id) {
    try {
      const response = await apiClient.delete(`/bakeryreviews/delete/${id}`);
      return response;
    } catch (error) {
      console.error(`Error deleting bakery review ${id}:`, error);
      throw error;
    }
  }

  // Product Reviews
  async getAllProductReviews() {
    try {
      const response = await apiClient.get('/productreviews', true);
      return response;
    } catch (error) {
      console.error('Error fetching product reviews:', error);
      throw error;
    }
  }

  async getProductReviewsByUser(userId) {
    try {
      const response = await apiClient.get(`/productreviews/user/${userId}`, true);
      return response;
    } catch (error) {
      console.error(`Error fetching product reviews for user ${userId}:`, error);
      throw error;
    }
  }
  
  async getProductReviewsByProduct(productId) {
    try {
      const response = await apiClient.get(`/productreviews/product/${productId}`, true);
      return response;
    } catch (error) {
      console.error(`Error fetching reviews for product ${productId}:`, error);
      throw error;
    }
  }

  async createProductReview(reviewData) {
    try {
      const response = await apiClient.post('/productreviews/create', reviewData);
      return response;
    } catch (error) {
      console.error('Error creating product review:', error);
      throw error;
    }
  }

  async updateProductReview(id, reviewData) {
    try {
      const response = await apiClient.patch(`/productreviews/update/${id}`, reviewData);
      return response;
    } catch (error) {
      console.error(`Error updating product review ${id}:`, error);
      throw error;
    }
  }

  async deleteProductReview(id) {
    try {
      const response = await apiClient.delete(`/productreviews/delete/${id}`);
      return response;
    } catch (error) {
      console.error(`Error deleting product review ${id}:`, error);
      throw error;
    }
  }

  // Combined review methods for user profile
  async getUserReviews(userId) {
    try {
      // Fetch both types of reviews in parallel
      const [bakeryReviewsResponse, productReviewsResponse] = await Promise.all([
        this.getBakeryReviewsByUser(userId),
        this.getProductReviewsByUser(userId)
      ]);
      
      return {
        bakeryReviews: bakeryReviewsResponse?.bakeryReviews || [],
        productReviews: productReviewsResponse?.productReviews || []
      };
    } catch (error) {
      console.error(`Error fetching user reviews for user ${userId}:`, error);
      throw error;
    }
  }

  // Fetch bakery and product data for reviews where names are missing
  async _enrichReviewsWithDetails(reviews) {
    const enrichedReviews = [...reviews];
    
    // Process the reviews that need enrichment
    const bakeryReviewsToEnrich = enrichedReviews
      .filter(review => review.type === 'bakery' && (!review.itemName || review.itemName === 'Unknown Bakery'))
      .map(review => ({ index: enrichedReviews.indexOf(review), id: review.itemId }));
      
    const productReviewsToEnrich = enrichedReviews
      .filter(review => review.type === 'product' && (!review.itemName || review.itemName === 'Unknown Product'))
      .map(review => ({ index: enrichedReviews.indexOf(review), id: review.itemId }));
    
    // Fetch the missing bakery data
    await Promise.all(
      bakeryReviewsToEnrich.map(async ({ index, id }) => {
        try {
          const bakeryResponse = await apiClient.get(`/bakeries/${id}`, true);
          if (bakeryResponse && bakeryResponse.name) {
            enrichedReviews[index].itemName = bakeryResponse.name;
          }
        } catch (error) {
          console.error(`Error fetching bakery ${id} details:`, error);
        }
      })
    );
    
    // Fetch the missing product data
    await Promise.all(
      productReviewsToEnrich.map(async ({ index, id }) => {
        try {
          const productResponse = await apiClient.get(`/products/${id}`, true);
          if (productResponse && productResponse.name) {
            enrichedReviews[index].itemName = productResponse.name;
            
            // If we have bakery info for the product, add that too
            if (productResponse.bakery && productResponse.bakery.name) {
              enrichedReviews[index].bakeryName = productResponse.bakery.name;
            } else if (productResponse.bakeryId) {
              try {
                const bakeryResponse = await apiClient.get(`/bakeries/${productResponse.bakeryId}`, true);
                if (bakeryResponse && bakeryResponse.name) {
                  enrichedReviews[index].bakeryName = bakeryResponse.name;
                }
              } catch (error) {
                console.error(`Error fetching bakery for product ${id}:`, error);
              }
            }
          }
        } catch (error) {
          console.error(`Error fetching product ${id} details:`, error);
        }
      })
    );
    
    return enrichedReviews;
  }

  // User review statistics
  async getUserReviewStats(userId) {
    try {
      const { bakeryReviews, productReviews } = await this.getUserReviews(userId);
      
      // Calculate statistics
      const totalReviews = bakeryReviews.length + productReviews.length;
      
      // Calculate average rating across all reviews
      let totalRating = 0;
      bakeryReviews.forEach(review => {
        totalRating += review.overallRating || 0;
      });
      
      productReviews.forEach(review => {
        totalRating += review.overallRating || 0;
      });
      
      const averageRating = totalReviews > 0 ? totalRating / totalReviews / 2 : 0; // Convert to 5-star scale
      
      // Get most recent review
      const allReviews = [...bakeryReviews, ...productReviews];
      allReviews.sort((a, b) => new Date(b.created_at) - new Date(a.created_at));
      const mostRecentReview = allReviews.length > 0 ? allReviews[0] : null;
      
      return {
        totalReviews,
        bakeryReviews: bakeryReviews.length,
        productReviews: productReviews.length,
        averageRating,
        mostRecentReview
      };
    } catch (error) {
      console.error(`Error calculating user review stats for user ${userId}:`, error);
      throw error;
    }
  }

  // Get recent reviews for a user (limit can be specified)
  async getUserRecentReviews(userId, limit = 5) {
    try {
      const { bakeryReviews, productReviews } = await this.getUserReviews(userId);
      
      // Combine and format reviews
      const allReviews = [
        ...bakeryReviews.map(review => ({
          id: review.id,
          type: 'bakery',
          itemId: review.bakeryId,
          itemName: review.bakery_name || 'Unknown Bakery',
          rating: review.overallRating / 2, // Convert to 5-star scale
          date: review.created_at,
          comment: review.review,
          rawReview: review // Keep the original review for reference
        })),
        ...productReviews.map(review => ({
          id: review.id,
          type: 'product',
          itemId: review.productId,
          itemName: review.product_name || 'Unknown Product',
          rating: review.overallRating / 2, // Convert to 5-star scale
          date: review.created_at,
          comment: review.review,
          rawReview: review // Keep the original review for reference
        }))
      ];
      
      // Sort by date (newest first)
      allReviews.sort((a, b) => new Date(b.date) - new Date(a.date));
      
      // Get the most recent reviews
      const recentReviews = allReviews.slice(0, limit);
      
      // Enrich the recent reviews with full details
      const enrichedReviews = await this._enrichReviewsWithDetails(recentReviews);
      
      return enrichedReviews;
    } catch (error) {
      console.error(`Error fetching recent reviews for user ${userId}:`, error);
      throw error;
    }
  }
  
  // Delete a review of either type
  async deleteReview(reviewId, reviewType) {
    try {
      if (reviewType === 'bakery') {
        return await this.deleteBakeryReview(reviewId);
      } else if (reviewType === 'product') {
        return await this.deleteProductReview(reviewId);
      } else {
        throw new Error(`Unknown review type: ${reviewType}`);
      }
    } catch (error) {
      console.error(`Error deleting ${reviewType} review ${reviewId}:`, error);
      throw error;
    }
  }
}

export default new ReviewService();